core/src/main/scala/org/apache/spark/sql/SnappyStrategies.scala

```diff
@@ -78,119 +79,142 @@ private[sql] trait SnappyStrategies {
-          if isCollocatedJoin(joinType, left, leftKeys, right, rightKeys) =>
-          val buildLeft = canBuildLeft(joinType) && canBuildLocalHashMap(left, conf)
-          if (buildLeft && left.statistics.sizeInBytes < right.statistics.sizeInBytes) {
+        case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right) =>
+          // check for hash join with replicated table first
+          if (canBuildRight(joinType) && canLocalJoin(right)) {
             makeLocalHashJoin(leftKeys, rightKeys, left, right, condition,
-              joinType, joins.BuildLeft, replicatedTableJoin = false)
-          } else if (canBuildRight(joinType) && canBuildLocalHashMap(right, conf)) {
-            makeLocalHashJoin(leftKeys, rightKeys, left, right, condition,
-              joinType, joins.BuildRight, replicatedTableJoin = false)
-          } else if (buildLeft) {
+              joinType, joins.BuildRight, replicatedTableJoin = true)
+          } else if (canBuildLeft(joinType) && canLocalJoin(left)) {
             makeLocalHashJoin(leftKeys, rightKeys, left, right, condition,
-              joinType, joins.BuildLeft, replicatedTableJoin = false)
-          } else if (RowOrdering.isOrderable(leftKeys)) {
-            joins.SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
-              planLater(left), planLater(right)) :: Nil
-          } else Nil
-
-        case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, condition, left, right)
-          if canBuildRight(joinType) && canBroadcast(right, conf) =>
-          if (skipBroadcastRight(joinType, left, right, conf)) {
+              joinType, joins.BuildLeft, replicatedTableJoin = true)
+          }
+          // check for collocated joins before going for broadcast
+          else if (isCollocatedJoin(joinType, left, leftKeys, right, rightKeys)) {
+            val buildLeft = canBuildLeft(joinType) && canBuildLocalHashMap(left, conf)
+            if (buildLeft && left.statistics.sizeInBytes < right.statistics.sizeInBytes) {
+              makeLocalHashJoin(leftKeys, rightKeys, left, right, condition,
+                joinType, joins.BuildLeft, replicatedTableJoin = false)
+            } else if (canBuildRight(joinType) && canBuildLocalHashMap(right, conf)) {
+              makeLocalHashJoin(leftKeys, rightKeys, left, right, condition,
+                joinType, joins.BuildRight, replicatedTableJoin = false)
+            } else if (buildLeft) {
+              makeLocalHashJoin(leftKeys, rightKeys, left, right, condition,
+                joinType, joins.BuildLeft, replicatedTableJoin = false)
+            } else if (RowOrdering.isOrderable(leftKeys)) {
+              joins.SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+                planLater(left), planLater(right)) :: Nil
+            } else Nil
+          }
+          // broadcast joins preferred over exchange+local hash join or SMJ
```

Comment:
Why is this reference of SMJ here, when in both if and else cases below hash join is being done.

Reply:
I did not see reference for SortMergeJoinExec other than line number 104.

Reply:
This is the case of "isCollocatedJoin" when the join is a collocated one but too large for a hash join. Otherwise it will return Nil both here and below if all custom join plans fail. In that case Spark's normal join (exchange+SMJ for equi-join and NLJ for non-equi joins) strategies will take effect which is applied after these strategies. Keeping the call for SMJ here for colocated case helps in avoiding an unnecessary exchange that Spark strategies may end up applying.

Reply:
Thanks Sumedh. I did not see reference for  Spark's normal join (exchange+SMJ). So is that default and comes into play if Nil is returned from here?

Reply:
Yes these strategies apply first, then SparkStrategies (JoinSelection in SparkStrategies.scala). See DefaultPlanner.strategies in SnappySessionState which adds the snappy strategies and then all the spark ones.